### PR TITLE
[SPARK-48329][SQL] Enable `spark.sql.sources.v2.bucketing.pushPartValues.enabled` by default

### PR DESCRIPTION
**Pull request by szehon-ho**
### What changes were proposed in this pull request?

This PR aims to enable `spark.sql.sources.v2.bucketing.pushPartValues.enabled` by default for Apache Spark 4.0.0 while keeping `spark.sql.sources.v2.bucketing.enabled` is `false`.

### Why are the changes needed?

`spark.sql.sources.v2.bucketing.pushPartValues.enabled` was added at Apache Spark 3.4.0 and has been used as one of the datasource v2 bucketing feature. This PR will help the datasource v2 bucketing users use this feature more easily. 

Note that this change is technically no-op for the default users because `spark.sql.sources.v2.bucketing.enabled` is `false` still.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Pass the CIs.

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

File: docs/sql-migration-guide.md [hunk 1]
```diff
@@ -55,6 +55,7 @@ license: |
...
 - Since Spark 4.0, A bug falsely allowing `!` instead of `NOT` when `!` is not a prefix operator has been fixed. Clauses such as `expr ! IN (...)`, `expr ! BETWEEN ...`, or `col ! NULL` now raise syntax errors. To restore the previous behavior, set `spark.sql.legacy.bangEqualsNot` to `true`. 
 - Since Spark 4.0, Views allow control over how they react to underlying query changes. By default views tolerate column type changes in the query and compensate with casts. To restore the previous behavior, allowing up-casts only, set `spark.sql.viewSchemaBindingMode` to `DISABLED`. This disables the feature and also disallows the `WITH SCHEMA` clause.
+- Since Spark 4.0, The Storage-Partitioned Join feature flag `spark.sql.sources.v2.bucketing.enabled` and `spark.sql.sources.v2.bucketing.pushPartValues.enabled` is set to `true`. To restore the previous behavior, set `spark.sql.sources.v2.bucketing.enabled` and `spark.sql.sources.v2.bucketing.pushPartValues.enabled` to `false`.
```
Comment (dongjoon-hyun) [inline docs/sql-migration-guide.md:58]: Please remove `The Storage-Partitioned Join feature flag`.

File: sql/core/src/test/scala/org/apache/spark/sql/connector/KeyGroupedPartitioningSuite.scala [hunk 2]
```diff
@@ -1169,7 +1169,6 @@ class KeyGroupedPartitioningSuite extends DistributionAndOrderingSuiteBase {
     Seq(true, false).foreach { shuffle =>
       withSQLConf(
         SQLConf.V2_BUCKETING_SHUFFLE_ENABLED.key -> shuffle.toString,
-        SQLConf.V2_BUCKETING_PUSH_PART_VALUES_ENABLED.key -> "true",
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/KeyGroupedPartitioningSuite.scala:1172]: Please revert all test case changes from this PR, @szehon-ho .

File: sql/core/src/test/scala/org/apache/spark/sql/execution/exchange/EnsureRequirementsSuite.scala [hunk 3]
```diff
@@ -1024,93 +1024,92 @@ class EnsureRequirementsSuite extends SharedSparkSession {
   }
 
   test("SPARK-41413: check compatibility when partition values mismatch") {
-    withSQLConf(SQLConf.V2_BUCKETING_PUSH_PART_VALUES_ENABLED.key -> "true") {
-      val leftPartValues = Seq(Array[Any](1, 1), Array[Any](2, 2)).map(new GenericInternalRow(_))
-      val rightPartValues = Seq(Array[Any](1, 1), Array[Any](2, 2), Array[Any](3, 3))
-          .map(new GenericInternalRow(_))
+    val leftPartValues = Seq(Array[Any](1, 1), Array[Any](2, 2)).map(new GenericInternalRow(_))
+    val rightPartValues = Seq(Array[Any](1, 1), Array[Any](2, 2), Array[Any](3, 3))
+        .map(new GenericInternalRow(_))
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/exchange/EnsureRequirementsSuite.scala:1029]: ditto. Please don't touch any test cases.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
The AS-IS PR title seems to be partial because this PR changes two configurations.

> [SPARK-48329][SQL] SPJ: Default spark.sql.sources.v2.bucketing.pushPartValues.enabled to true 

Also, please resolve the conflicts and revert the test coverage change, @szehon-ho .

**Comment by dongjoon-hyun** [comment]
cc @viirya , @sunchao

File: docs/sql-migration-guide.md (hunk 1 above)
Comment (szehon-ho) [reply docs/sql-migration-guide.md:60]: Ok I took this from @superdiaodiao , who also worked on this pr, will revert.

File: sql/core/src/test/scala/org/apache/spark/sql/connector/KeyGroupedPartitioningSuite.scala (hunk 2 above)
Comment (szehon-ho) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/KeyGroupedPartitioningSuite.scala:1172]: OK

**Comment by szehon-ho** [comment]
Thanks @dongjoon-hyun sorry the pr was not ready.  I was trying to integrate the changes from @superdiaodiao who I asaw also made a pr for the same, so we can be co-authors.  Reverted the additional config change and test case, will check the test result.

**Comment by superdiaodiao** [comment]
> Thanks @dongjoon-hyun sorry the pr was not ready.  I was trying to integrate the changes from @superdiaodiao who I asaw also made a pr for the same, so we can be co-authors.  Reverted the additional config change and test case, will check the test result.

I will close my PR and the rest work needs you, let's co-author this time.

**Comment by superdiaodiao** [comment]
> Thanks @dongjoon-hyun sorry the pr was not ready. I was trying to integrate the changes from @superdiaodiao who I asaw also made a pr for the same, so we can be co-authors. Reverted the additional config change and test case, will check the test result.

I will close my PR and you can continue, let's co-author this time.

**Comment by szehon-ho** [comment]
Thanks!  @dongjoon-hyun @sunchao  can you take another look?

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @szehon-ho and @superdiaodiao 

Merged to master for Apache Spark 4.0.0.

**Comment by dongjoon-hyun** [comment]
BTW, @szehon-ho and @superdiaodiao .

The umbrella JIRA is closed one already for Apache Spark 3.3.0. I moved SPARK-48329 to a subtask of SPARK-44111 for now.

Screenshot 2024-05-21 at 10 02 39

For the other new tasks, please create a new umbrella JIRA and use it.

**Comment by superdiaodiao** [comment]
> BTW, @szehon-ho and @superdiaodiao .
> 
> The umbrella JIRA is closed one already for Apache Spark 3.3.0. I moved SPARK-48329 to a subtask of SPARK-44111 for now.
> 
> Screenshot 2024-05-21 at 10 02 39
> 
> For the other new tasks, please create a new umbrella JIRA and use it.

OK

**Comment by szehon-ho** [comment]
@dongjoon-hyun do you have any guidance how we may have a new Spark 4.0+ tracking JIRA to link all the new SPJ items?  I think it will be nice to have all of them in one list.

**Comment by dongjoon-hyun** [comment]
Feel free to create a new umbrella Jira issue with a proper meaningful title instead of duplicating old Jira issue title. Then, link it to SPARK-44111, @szehon-ho . That's enough.

**Pull request closed without merging**